src/inventory_ui.cpp

```diff
@@ -288,7 +316,18 @@ inventory_selector_preset::inventory_selector_preset()
 bool inventory_selector_preset::sort_compare( const inventory_entry &lhs,
         const inventory_entry &rhs ) const
 {
-    return lhs.cached_name.compare( rhs.cached_name ) < 0; // Simple alphabetic order
+    if( lhs.cached_name == rhs.cached_name ) {
+        std::string const ltname = lhs.any_item()->tname( 1, false );
+        std::string const rtname = rhs.any_item()->tname( 1, false );
+        if( ltname == rtname ) {
+            // FIXME: find a better way to sort identically named items
+            // this is needed for sorting contents below containers
+            return lhs.any_item().get_item() < rhs.any_item().get_item();
+        }
+        return localized_compare( ltname, rtname );
+    }
+
+    return localized_compare( lhs.cached_name, rhs.cached_name );
```

Comment:
You can write this as something like
```cpp
auto sort_key = []( const inventory_entry &e ) {
    return std::make_tuple( e.cached_name, e.any_item()->tname( 1, false ) );
}

return localized_compare( sort_key( lhs ), sort_key( rhs ) );
```

Please don't compare pointer values (the return values of `get_item`).  That makes the behaviour non-deterministic, which can be confusing for users and annoying for unit tests.  If there's no reason for one item to compare less than the other, that's fine; we can have a strict weak ordering rather than a total ordering.  But you might want to use `stable_sort` rather than `sort`.

Or maybe there's some reason we can't have non-comparable items.  I don't currently understand the details of what you've done below.

Reply:
Without strict ordering I get
<details>
<summary>this</summary>

![this](https://user-images.githubusercontent.com/68240139/145342476-fea71c81-0956-495b-8c03-a011a14cbbd2.png)
</details>
instead of
<details>
<summary>this</summary>

![strict ordering](https://user-images.githubusercontent.com/68240139/145342576-d16b5fa9-6c9d-4c37-8114-bdd5dde04340.png)
</details>
(4th test case in my PR description)

Reply:
I've pushed a commit with your suggestion, but I still need the pointer comparison. The old `order_by_parent()` also compared pointers but masked it with a `reinterpret_cast` (and I've done the same here to show it's intentional).

I'd be very happy to try other ideas though.

Reply:
I've added a sort of unique ID to items as it wasn't all that much work. Let me know if there are further issues (and sorry for the many pings)